src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/nd4jutils/Nd4jIOUtils.java

```diff
@@ -0,0 +1,161 @@
+
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.tsv.DataLine;
+import org.broadinstitute.hellbender.utils.tsv.TableColumnCollection;
+import org.broadinstitute.hellbender.utils.tsv.TableReader;
+import org.broadinstitute.hellbender.utils.tsv.TableWriter;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.IntStream;
+
+/**
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class Nd4jIOUtils {
+
+    /**
+     * Write NDArray to a binary dump
+     *
+     * @param arr an instance of {@link INDArray}
+     * @param outputFile the output path
+     */
+    public static void writeNDArrayToBinaryDumpFile(@Nonnull final INDArray arr,
+                                                    @Nonnull final File outputFile) {
+        try (final DataOutputStream dos = new DataOutputStream(new FileOutputStream(outputFile))) {
+            Nd4j.write(arr, dos);
+        } catch (final IOException ex) {
+            throw new UserException.CouldNotCreateOutputFile(outputFile, "Could not write the NDArray");
+        }
+    }
+
+    /**
+     * Read NDArray from a binary dump
+     *
+     * @param inputFile the input Nd4j binary dump file
+     * @return an {@link INDArray}
+     */
+    public static INDArray readNDArrayFromBinaryDumpFile(@Nonnull final File inputFile) {
+        try (final DataInputStream dis = new DataInputStream(new FileInputStream(inputFile))) {
+            return Nd4j.read(dis);
+        } catch (final IOException ex) {
+            throw new UserException.CouldNotReadInputFile(inputFile, "Could not read the input file");
+        }
+    }
+
+    /**
+     * Write NDArray to a tab-separated file
+     *
+     * @param arr an instance of {@link INDArray}
+     * @param outputFile the output file
+     * @param columnNames column names
+     * @param rowNames row names
+     */
+    public static void writeNDArrayToTextFile(@Nonnull final INDArray arr,
+                                              @Nonnull final File outputFile,
+                                              @Nullable final List<String> rowNames,
+                                              @Nullable final List<String> columnNames) {
+        Utils.validateArg(arr.rank() <= 2, "only rank-1 and rank-2 NDArray objects can be saved");
+
+        final int rowDimension = arr.shape()[0];
+        final int colDimension = arr.shape()[1];
+
+        final TableColumnCollection columnNameCollection;
+        final List<String> columnNameList = new ArrayList<>();
+        if (columnNames == null) {
+            columnNameList.add("ROW_NAME");
```

Comment:
This is addressed in the post-PR commit https://github.com/broadinstitute/gatk-protected/pull/857/commits/4891f92965877c54c20d92835070476da4eefa8f